dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -1697,6 +1697,8 @@ RtpsUdpDataLink::received(const RTPS::AckNackSubmessage& acknack,
     callbacks[i]->reader_exists(remote, local);
   }
 
+  acknowledged_by_reader(local, remote);
```

Comment:
If this gets called every time we get an ACK, then are we guaranteed to always send the keys twice? Should we be storing the sequence numbers used when we sent the keys the first time and comparing against the sequence numbers sent in the ACK so we only resend when necessary?

Reply:
So this is starting to get a bit hazy on why exactly this is set up how it is, but I think that the key sender has no way to know if the key exchange was successful, because the volatile reader accepts the sequence number as is when it first encounters it. Resending the keys is a waste of resources, but it's the only way to ensure the keys are being exchanged short of some sort of fundamental change in the key exchange process works.

Reply:
Ah, right, sorry, I'd forgotten the underlying issue. If we successfully got all the keys we'd ACK the final key, and if we'd never gotten any of them but saw that the latest heartbeat was past the final key, we'd ACK the final key in that situation too since that's the first data point we'd know about. I suppose if we wanted to avoid the wastefulness we could only ever "resend" the keys, but then there's additional delay waiting for an ACKNACK, right? Maybe for OpenDDS we can send additional heartbeats on the volatile reader? Could we even send additional "initial" heartbeats (high SN is set to zero?) when we want to send keys to a reader to prompt the faster ACKNACK reply?